### PR TITLE
Fix sessionless clients

### LINKED ISSUE #25059: Consolidate device token with session
When using clients like cadaver, no cookies are used. So even with a device token, it will generate a new authentication every time.

One idea would be to find a way to make the same device token reuse the same session.
Not sure about security implications though.

As discussedon IRC @ChristophWurst @DeepDiver1975 @guruz @rullzer

### PR DESCRIPTION
**Pull request by ChristophWurst**
fixes #25059

TODO
- [x] fix/add tests

This way we do not create a session token for clients if they do not support cookies. Otherwise the list of connected browsers was spammed.

Can be tested with curl:

``` bash
$ for x in {1..50}; do curl --request PROPFIND -u admin:YDOAF-MFAGV-LXZEE-JTERS http://localhost:8080/remote.php/webdav; done
```

cc @LukasReschke @PVince81 @guruz

### CODE REVIEW

File: lib/private/User/Session.php [hunk 1]
```diff
@@ -368,9 +369,22 @@ public function logClientIn($user, $password) {
...
+		if ($this->supportsCookies($request)) {
+			$this->createSessionToken($request, $this->getUser()->getUID(), $user, $password);
+		}
+
 		return true;
 	}
 
+	protected function supportsCookies(IRequest $request) {
+		if (!is_null($request->getCookie('cookie_test'))) {
+			return true;
```
Comment (ChristophWurst) [inline lib/private/User/Session.php]: we could unset the cookie here so the client doesn't need to resend it

File: lib/private/User/Session.php [hunk 2]
```diff
@@ -428,7 +442,8 @@ protected function prepareUserLogin() {
 	 */
 	public function tryBasicAuthLogin(IRequest $request) {
 		if (!empty($request->server['PHP_AUTH_USER']) && !empty($request->server['PHP_AUTH_PW'])) {
-			$result = $this->logClientIn($request->server['PHP_AUTH_USER'], $request->server['PHP_AUTH_PW']);
+			$request = \OC::$server->getRequest();
```
Comment (nickvergessen) [inline lib/private/User/Session.php]: overwriting he parameter?

Comment (ChristophWurst) [inline lib/private/User/Session.php]: makes no sense, I should use the parameter directly instead

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -368,9 +369,22 @@ public function logClientIn($user, $password) {
...
+		}
+
 		return true;
 	}
 
+	protected function supportsCookies(IRequest $request) {
+		if (!is_null($request->getCookie('cookie_test'))) {
+			return true;
+		}
+		setcookie('cookie_test', 'test', $this->timeFacory->getTime() + 3600);
```
Comment (LukasReschke) [inline lib/private/User/Session.php]: I'd give that a longer lifetime. Doesn't hurt to have it living unlimited.

File: lib/private/User/Session.php (hunk 1 above)
Comment (LukasReschke) [inline lib/private/User/Session.php]: I'd keep it for simplicity.

**Comment by LukasReschke** [comment]
👍

File: lib/private/User/Session.php (hunk 3 above)
Comment (PVince81) [inline lib/private/User/Session.php]: `timeFacory`

You might be able to use the app framework's response object to set the cookie, this way it's mockable and you can write a tets for it.

Comment (ChristophWurst) [inline lib/private/User/Session.php]: well, we create the session token only to reduce the user backend load. If the time between two requests is greater than 5mins we actually don't need to create a session token as the password is then checked on both requests anyway.

**Comment by PVince81** [comment]
From what I understand, when setting the `cookie_test`, the check will pass when the second request from that client comes in and only at this point the token is created.
Now what happens if more than one request arrives as the second request ? (two parallel "second requests, race condition). Will this still work properly or will two tokens be created ?

**Comment by ChristophWurst** [comment]
> Will this still work properly or will two tokens be created ?

As discussed yesterday this is still an issue. Since fixing this will need another refactoring of the auth system I'd like to postpone this to 9.2.

**Comment by PVince81** [comment]
If we consider that having two session tokens is ok, can you elaborate whether this is still running properly ? Would this cause any disturbance in the client ?

**Comment by ChristophWurst** [comment]
It is. It just means that an old, unused token will exist in the database.

**Comment by ChristophWurst** [comment]
(…which will be cleaned up after 1d of inactivity)

**Comment by PVince81** [comment]
Fair enough :+1:

**Comment by PVince81** [comment]
CC @owncloud/qa

**Pull request merged**